libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+    constexpr _CUDA_VSTD::size_t __copy_size = 1;
+    using __chunk_t = __copy_chunk<__copy_size>;
+
+    // "Group"-strided loop over memory
+    const size_t __stride = __g.size() * __copy_size;
+
+    for (_CUDA_VSTD::size_t __offset = __g.thread_rank() * __copy_size; __offset < __size; __offset += __stride) {
+        __chunk_t tmp = *reinterpret_cast<const __chunk_t *>(__src + __offset);
+        *reinterpret_cast<__chunk_t *>(__dest + __offset) = tmp;
+    }
 }
 
-template<_CUDA_VSTD::size_t _Native_alignment, typename _Group, typename _Sync>
-_LIBCUDACXX_INLINE_VISIBILITY
-async_contract_fulfillment inline __memcpy_async(
-        _Group const & __group, char * __destination, char const * __source,
-        _CUDA_VSTD::size_t __size, _Sync & __sync) {
-    async_contract_fulfillment __is_async = async_contract_fulfillment::none;
+/***********************************************************************
+ * cuda::memcpy_async dispatch helper functions
+ *
+ * - __get_size_align struct to determine the alignment from a size type.
+ ***********************************************************************/
 
-    NV_DISPATCH_TARGET(
-        NV_PROVIDES_SM_80,
-
-        if (__isShared(__destination) && __isGlobal(__source)) {
-            if (_Native_alignment < 4) {
-                auto __source_address = reinterpret_cast<_CUDA_VSTD::uintptr_t>(__source);
-                auto __destination_address = reinterpret_cast<_CUDA_VSTD::uintptr_t>(__destination);
-
-                // Lowest bit set will tell us what the common alignment of the three values is.
-                auto _Alignment = __ffs(__source_address | __destination_address | __size);
-
-                switch (_Alignment) {
-                    default:
-                        __memcpy_async_sm_dispatch<_Native_alignment>(__group, __destination, __source, __size, __sync, __is_async);
-                    case 4: __is_async = __memcpy_async_impl<8>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 3: __is_async = __memcpy_async_impl<4>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                    case 2: // fallthrough
-                    case 1: __is_async = __memcpy_async_impl<1>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size()); break;
-                }
-            }
-            else {
-                __is_async = __memcpy_async_impl<_Native_alignment>::__copy(__destination, __source, __size, __group.thread_rank(), __group.size());
+// The __get_size_align struct provides a way to interrogate the guaranteed
+// "alignment" of a provided size. In this case, an n-byte aligned size means
+// that the size is a multiple of n.
+//
+// Use as follows:
+// static_assert(__get_size_align<size_t>::align == 1)
+// static_assert(__get_size_align<aligned_size_t<n>>::align == n)
+template< class... >
+using __custom_void_t = void;
```

Comment:
Nitpick:

This chould be `_CUDA_VSTD::__voit_t` when you add `#include "../type_traits/void_t.h"`

Reply:
Thanks! was looking for a way to use a predefined type.

Changed now.